### ISSUE TITLE
Client-side validation for RPC parameter types

### ISSUE THREAD

**Issue by vlukashov** [vaadin/vaadin-connect#213]
When showing validation error messages in the UI, I want to run some validation rules on the client so that I can show the error messages quicker and not rely on network availability for form validation.

depends on #169 

### In scope:
- There is a client-side validation error if a given JS value is obviously not in a format compatible with the expected Java type. For every supported Java type there is at least some JS values that would be rejected by client-side validation.
- Client-side validation runs automatically before any RPC call (and the returned promise is rejected in case of a validation error)
- There is a client-side API to run validation without making an RPC call (two modes are available: client-only and client-server)

### Out of scope:
- Most common JSR-380 constraints are validated on the client (without a round trip): `@NotNull`, `@NotBlank`, `@Required`, `@Min`, `@Max`
- All built-in JSR-380 (Bean Validation 2.0) constraints are validated on the client (without a round trip)
- JSR-380 validation groups are supported (i.e. in JSR-380 there is a way to declare certain constraints as relevant only in certain contexts; In Vaadin Connect it is possible to define which contexts are relevant: per-method / class / globally)
- JSR-380 extension constraints are validated on the server (with a round trip). This includes cross-field validation
- It’s possible to add JS implementations for JSR-380 extension constraints to avoid a round trip for validation
- No sync API for client-only validation (always returns a promise for consistency)
- No UI updates in response to validation failures.
- No specific support for internationalization
